### ISSUE TITLE
Empty string literal "" matches `Sequence[Never]` overload instead of `str`

### ISSUE THREAD

**Issue by cmp0xff** [astral-sh/ty#4365]
### Summary

When an empty string literal `""` is passed to an overloaded function, `ty` matches it against an earlier `Sequence[Never]` overload (treating `""` as an empty sequence of characters) instead of matching the dedicated `str` overload.

### Minimal Example

https://play.ty.dev/ab067b55-b902-4b09-b21d-8650640e8982

```py
from typing import Never, Sequence, assert_type, overload

@overload
def create_container(data: Sequence[Never]) -> list[object]: ...
@overload
def create_container(data: Sequence[str]) -> list[str]: ...
def create_container(data: object) -> object:
    return data

# Non-empty string: PASSES in ty (inferred as list[str])
assert_type(create_container("hello"), list[str])

# Empty string: FAILS in ty (inferred as list[object] via Sequence[Never])
assert_type(create_container(""), list[str])
```

### Diagnostic

```text
error[type-assertion-failure]: Argument does not have asserted type `str`
  --> test.py:13:1
   |
13 | assert_type(create_container(""), str)
   | ^^^^^^^^^^^^--------------------^^^^^^
   |             |
   |             Inferred type is `list[object]`
info: `str` and `list[object]` are not equivalent types
```

### Expected Behaviour

While str implements Sequence[str], passing a string literal (including "") to an overloaded function should prioritize the exact str parameter over generic sequence overloads like Sequence[object] / Sequence[Never].

**Comment by AlexWaygood** [comment]
Thanks for the report!

> ### Expected Behaviour
> While str implements Sequence[str], passing a string literal (including "") to an overloaded function should prioritize the exact str parameter over generic sequence overloads like Sequence[object] / Sequence[Never].

Can you say a bit more about why you'd expect this behaviour? Our current behaviour seems correct to me. `""` is indeed an empty sequence.

**Comment by cmp0xff** [comment]
Hi, thank you for triaging.

Both may be correct, so this is more a preference. All other type checkers give string instead of Sequence[Never]. https://pyrefly.org/sandbox/?project=v2.w1tk4ihFEMUDloIVqTSCFovwdK2RYwXXGw2WioWlN0hcAA3UAHpOCRh9mKGrrBAAEVCAOhToPGC0pJcCSy-gC_Vi8pSBpqcqZJSUFBRb6evDtOcXpeun5umn5CcX66Po0FcAJaTEssRMcJmIapxCPjg1FAONpSRa1SGxqo4vUtVBcQpUgTeeAA https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUkBHAVySxiQChRIpUBPCuiwBzMpWq0QSAG5IC8APoCKSADRQAcrKQhOnACZJgUYAAp4ALnpNW7JAG1tckAF0AlJc5QfUCgRwcfXMAIhD3TiA https://mypy-play.net/?gist=a1057196c8e495ee29339f3ccc69fdfa

In pandas-stubs, we would like to have Series("") giving a Series[str], instead of Series[Any]. This was the motivation.

**Comment by AlexWaygood** [comment]
The reason why all other type checkers select the second overload is that no other type checker understand that `Literal[""]` is a subtype of `Sequence[Never]`, which leads to false-positive diagnostics on things like this:

```py
from typing import Sequence, Never

def f(x: Sequence[Never]): ...

f("")
```

So I think this is a bug in all other type checkers rather than a bug in ty

**Comment by carljm** [comment]
@cmp0xff Why does it make sense to have an overload accepting `Sequence[Never]` here in the first place? Is that a common type people pass in?

**Comment by cmp0xff** [comment]
> [@cmp0xff](https://github.com/cmp0xff) Why does it make sense to have an overload accepting `Sequence[Never]` here in the first place? Is that a common type people pass in?

Hi @carljm , we need a pattern for `Sequence[Any]` as well as for an empty `Sequence` as a guard.

For example, there can be a few `__new__` overloads for `Series`, and the first "normal" overload may go to `Sequence[bool]`, giving `Series[bool]`. However we also need to have an overload when `Sequence` is `Any` or `Unknown`, and we need to get `Series[Any]`.

If we don't have the guard and it comes with `Sequence[Any]`, we will get the first matched case `Sequence[bool]`, giving us `Series[bool]`, which is not what we want.

It is a common design pattern we use in `pandas-stubs`.

**Comment by carljm** [comment]
@cmp0xff I see. This is unfortunate. I don't think this is a good pattern for expressing overload behavior, but I think that the poor behavior of overload resolution in current Python type checkers has left you without better options. 

This pattern is a workaround for two things:

1. The poor behavior specified in the Python typing spec, where ambiguous overloads resolve to `Any` rather than picking the most precise common type between the matched ambiguous overloads. That is, if the input type is `Sequence[Any]` and it matches both `Sequence[int]` and `Sequence[str]`, and those overloads would return `Series[int]` and `Series[str]`, a better type checker behavior would be to automatically retain the common `Series` portion of the type and infer `Series[Any]`, rather than inferring just `Any`.
2. The poor behavior of other type checkers (not following the spec) where they sometimes resolve gradual ambiguous overloads by just arbitrarily picking the first match, rather than by following the spec and inferring `Any`.

It seems to me that if these two things were fixed, you wouldn't need these confusing `Never` overloads.

ty already doesn't do the second, and I would like to improve our behavior on the first. Unfortunately this won't really help, since you also need to support other type checkers.

**Comment by carljm** [comment]
Hoping to make some positive progress on this issue at https://discuss.python.org/t/permitting-more-precise-inferred-types-from-ambiguous-overloads/108776
